### ISSUE TITLE
Failed to fetch / http: proxy error: context canceled / Data proxy error

### ISSUE THREAD

**Issue by erkexzcx** [grafana/grafana#24611]
**What happened**:
Grafana randomly doesn't show data when loading/refreshing dashboard. Go through added screenshots first to understand the issue.

**What you expected to happen**:
Grafana to load panels as expected.

**How to reproduce it (as minimally and precisely as possible)**:
Load or refresh dashboard.

**Anything else we need to know?**:
Grafana is under Nginx proxy.

It looks like only one of dashboards is affected, but I managed to get this error in another dashboard only once, but I believe this is something broken with Grafana. There are no errors in Nginx (both _access.log_ and _error.log_), no errors in InfluxDB as well.

Taking URL from browser's developers tools and opening it directly, JSON data is always returned successfully without any errors or cancelled/failed web requests, so I do think there is some sort of DOS prevention in Grafana..? Or Grafana is not able to handle _that many_ requests at once?

Here are some logs from `/var/log/grafana/grafana.log`:
```
t=2020-05-13T09:19:39+0000 lvl=eror msg="Data proxy error" logger=data-proxy-log userId=2 orgId=1 uname=name.surname path=/api/datasources/proxy/11/query remote_addr=127.0.0.1 referer="https://grafana.example.com/d/xxxxxxxxx/dashboard-title?orgId=1&refresh=5m&from=now-1h&to=now" error="http: proxy error: context canceled"
t=2020-05-13T09:19:39+0000 lvl=eror msg="Data proxy error" logger=data-proxy-log userId=2 orgId=1 uname=name.surname path=/api/datasources/proxy/11/query remote_addr=127.0.0.1 referer="https://grafana.example.com/d/xxxxxxxxx/dashboard-title?orgId=1&refresh=5m&from=now-1h&to=now" error="http: proxy error: context canceled"
t=2020-05-13T09:19:39+0000 lvl=eror msg="Data proxy error" logger=data-proxy-log userId=2 orgId=1 uname=name.surname path=/api/datasources/proxy/11/query remote_addr=127.0.0.1 referer="https://grafana.example.com/d/xxxxxxxxx/dashboard-title?orgId=1&refresh=5m&from=now-1h&to=now" error="http: proxy error: context canceled"
t=2020-05-13T09:19:39+0000 lvl=eror msg="Data proxy error" logger=data-proxy-log userId=2 orgId=1 uname=name.surname path=/api/datasources/proxy/11/query remote_addr=127.0.0.1 referer="https://grafana.example.com/d/xxxxxxxxx/dashboard-title?orgId=1&refresh=5m&from=now-1h&to=now" error="http: proxy error: context canceled"
t=2020-05-13T09:19:39+0000 lvl=eror msg="Data proxy error" logger=data-proxy-log userId=2 orgId=1 uname=name.surname path=/api/datasources/proxy/11/query remote_addr=127.0.0.1 referer="https://grafana.example.com/d/xxxxxxxxx/dashboard-title?orgId=1&refresh=5m&from=now-1h&to=now" error="http: proxy error: context canceled"

```

**Environment**:
- Grafana version: Grafana v6.7.3 (a04ef6cefc)
- Data source type & version: InfluxDB (latest stable), 1.8.0
- OS Grafana is installed on: Debian 10.
- User OS & Browser: Arch Linux, Google Chrome (this error is not just me, but for other users as well)
- Grafana plugins: Only default plugins in use
- Others: There is duplicate (or very similar issue) without any solutions, raised few years back: https://github.com/grafana/grafana/issues/11069

**Screenshots**:

Nothing loaded this time:
![image](https://user-images.githubusercontent.com/5955795/81792291-8add5d80-9510-11ea-8576-1f966b31f5c1.png)

After a refresh, there is _some_ panels showing data:
![image](https://user-images.githubusercontent.com/5955795/81792598-ef98b800-9510-11ea-867c-5d767da31333.png)

Howering mouse over red triangles do not show anything useful:
![image](https://user-images.githubusercontent.com/5955795/81792705-148d2b00-9511-11ea-9ad5-2bf13be2d960.png)

Looking at the developer console, some web requests do not complete, get cancelled, or succeeds:
![image](https://user-images.githubusercontent.com/5955795/81792992-72ba0e00-9511-11ea-84be-f252863f5881.png)

Any assistance would be appreciated.

**Comment by ivanahuckova** [comment]
Hello @erkexzcx, could you try to access the database without Nginx in the middle and see if you still get errors?

**Comment by erkexzcx** [comment]
Could it be that this is the same issue? https://github.com/grafana/metrictank/issues/1820

I am not able to check without Nginx at the moment as it would cause downtime. Thing is - it works just fine when loading query using `curl` or with browser directly. No matter how fast I perform HTTP requests. I believe it's a problem when Grafana receives many different requests at once and tries to retrieve and send data to browser. I don't see how Nginx could be a point of failure...

Is it possible to enable more debugging logs or more meaningful messages? Setting logs `level = debug` does not provide anything more useful.

**Comment by erkexzcx** [comment]
Nevermind, issue solved. What happened is that we created variable in Grafana that dynamically creates list of entries. And in **just a single panel** we set "`variable` = `^$variable$`" and yeah, the final URL that browser tries to perform ended up being **6676 chars long**.

Not sure if Grafana, Nginx or Influxdb did not want to respond to this long URL, and all other requests that were made at the same time were also _failed_.

**Comment by erkexzcx** [comment]
@ivanahuckova Would be great to have some client side checks for max URL length that Grafana generated and is going to try to perform. And of course - some more meaningful errors.

**Comment by btai24** [comment]
I am running into this issue on grafana v7.0.3 using a community built dashboard. I was able to deploy it successfully on a test cluster with 3 nodes but once I tried to deploy this to another larger cluster (15 nodes), these same issues are persisting only on the larger cluster.

<img width="1593" alt="Screen Shot 2020-09-22 at 3 17 51 PM" src="https://user-images.githubusercontent.com/27699163/93952105-e4df0280-fcfc-11ea-91ed-2d8893ea60cd.png">

<img width="1649" alt="Screen Shot 2020-09-22 at 5 15 17 PM" src="https://user-images.githubusercontent.com/27699163/93952120-ead4e380-fcfc-11ea-9cfb-3537ca6d2b24.png">

I don't believe my issue is related to max URL length because the same query that succeeds in the smaller cluster will fail in the larger cluster.

Smaller cluster:
<img width="1645" alt="Screen Shot 2020-09-22 at 5 58 25 PM" src="https://user-images.githubusercontent.com/27699163/93952445-b4e42f00-fcfd-11ea-8195-99bcd2a73ff3.png">

Larger cluster:
<img width="1645" alt="Screen Shot 2020-09-22 at 5 57 36 PM" src="https://user-images.githubusercontent.com/27699163/93952456-bb72a680-fcfd-11ea-874a-abee78221c6c.png">

Datasource logs:
```
t=2020-09-23T00:39:42+0000 lvl=eror msg="Data proxy error" logger=data-proxy-log userId=2 orgId=1 uname=btai path=/api/datasources/proxy/1/api/v1/query_range remote_addr=127.0.0.1 referer="https://monitoring.<redacted>.com/d/hb7fSE0Zz/1-node-exporter-for-prometheus-dashboard-en-v20200628?orgId=1" error="http: proxy error: context canceled"
t=2020-09-23T00:39:42+0000 lvl=eror msg="Data proxy error" logger=data-proxy-log userId=2 orgId=1 uname=btai path=/api/datasources/proxy/1/api/v1/query_range remote_addr=127.0.0.1 referer="https://monitoring.<redacted>.com/d/hb7fSE0Zz/1-node-exporter-for-prometheus-dashboard-en-v20200628?orgId=1" error="http: proxy error: context canceled"
t=2020-09-23T00:39:42+0000 lvl=info msg="Request Completed" logger=context userId=2 orgId=1 uname=btai method=GET path=/api/datasources/proxy/1/api/v1/query_range status=502 remote_addr=127.0.0.1 time_ms=91 size=0 referer="https://monitoring.<redacted>.com/d/hb7fSE0Zz/1-node-exporter-for-prometheus-dashboard-en-v20200628?orgId=1"
t=2020-09-23T00:39:42+0000 lvl=info msg="Request Completed" logger=context userId=2 orgId=1 uname=btai method=GET path=/api/datasources/proxy/1/api/v1/query_range status=502 remote_addr=127.0.0.1 time_ms=91 size=0 referer="https://monitoring.<redacted>.com/d/hb7fSE0Zz/1-node-exporter-for-prometheus-dashboard-en-v20200628?orgId=1"
t=2020-09-23T00:39:42+0000 lvl=eror msg="Data proxy error" logger=data-proxy-log userId=2 orgId=1 uname=btai path=/api/datasources/proxy/1/api/v1/query_range remote_addr=127.0.0.1 referer="https://monitoring.<redacted>.com/d/hb7fSE0Zz/1-node-exporter-for-prometheus-dashboard-en-v20200628?orgId=1" error="http: proxy error: context canceled"
t=2020-09-23T00:39:42+0000 lvl=eror msg="Data proxy error" logger=data-proxy-log userId=2 orgId=1 uname=btai path=/api/datasources/proxy/1/api/v1/query_range remote_addr=127.0.0.1 referer="https://monitoring.<redacted>.com/d/hb7fSE0Zz/1-node-exporter-for-prometheus-dashboard-en-v20200628?orgId=1" error="http: proxy error: context canceled"
```

Extra info:
Grafana version: v7.0.3
Grafana datasource: Thanos v0.14.0
Kubernetes version: v1.17 
Deployed: prometheus-operator

**Comment by ivanahuckova** [comment]
Hello @btai24, can you please check if requests fail without proxy?

**Comment by btai24** [comment]
@ivanahuckova what do you mean by without proxy?

I spun up an alpine debug image in my cluster and tried a couple of the failed requests and they succeeded.

```
/ # curl -H "cookie: grafana_session=<redacted>" "prometheus-operator-grafana.monitoring/api/datasources
/proxy/1/api/v1/query?query=sum(node_memory_MemTotal_bytes%7Binstance%3D~%22172%5C%5C.25%5C%5C.108%5C%5C.159%3A9100%22%7D)&tim
e=1600884087"
{"status":"success","data":{"resultType":"vector","result":[{"metric":{},"value":[1600884087,"64393211904"]}]}}
/ # curl -H "cookie: grafana_session=<redacted>" "prometheus-operator-grafana.monitoring/api/datasources
/proxy/1/api/v1/query?query=sum(node_memory_MemTotal_bytes%7Binstance%3D~%22172%5C%5C.25%5C%5C.108%5C%5C.159%3A9100%22%7D)&tim
e=1600884087"
{"status":"success","data":{"resultType":"vector","result":[{"metric":{},"value":[1600884087,"64393211904"]}]}}```

**Comment by ivanahuckova** [comment]
Can you try if configuring Prometheus data source to use POST instead of GET solves our problem?

**Comment by lakshmitumati** [comment]
Hi! I am having this same issue while working with Loki in the latest version of Grafana. While uploading lots of data into an InfluxDB database, these errors showed up in the logs:
<img width="1333" alt="Screen Shot 2021-02-11 at 2 29 09 PM" src="https://user-images.githubusercontent.com/69651904/107688314-84ce0600-6c75-11eb-8f06-4c3b216356c9.png">

Do you know what these errors mean or why it may have caused Grafana to crash? Thanks!

**Comment by ivanahuckova** [comment]
@lakshmitumati Could you please create separate issue for this and provide us with more information about your problem. Thanks!
